### DISCUSSION TITLE
How to make custom control that accepts user provided content?

### DISCUSSION THREAD

**Discussion by WarWithinMe** [AvaloniaUI/Avalonia#11462] [Q&A] [answered]
I want to create a toolbar that has predefined buttons (red area in attached image) and allows the user of the toolbar to insert whatever content they want into a specific region (blue area). :

![image](https://github.com/AvaloniaUI/Avalonia/assets/567110/84680ec2-3b67-4617-be2b-70337a207e0a)

I would like the user to use it in such way:
```
<CustomToolbar>
    <StackPanel>
        <Button/>
        <Button/>
    </StackPanel>
</CustomToolbar>
```

Can someone kindly provide some hints on how to implement CustomToolbar in Avalonia? Or is it not a recommended way to make custom controls? I've checked several tutorials and am still not sure how to do it. Thanks in advance.

**Comment by maxkatz6** [comment, marked as answer]
ContentControl is ready to use example of one. You can inherit it to inherit the same functionality.
You can see it has Content and optional ContentTemplate - https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Controls/ContentControl.cs#L51-L66
Where Content can accept either Controls (like StackPanel in your case) or data objects. In case of data objects it requires ContentTemplate to correctly display that data.

Note, for your custom type you would need to add a style with template as well. As default one from ContentControl won't be inherited.
https://github.com/AvaloniaUI/Avalonia/blob/0.10.21/src/Avalonia.Themes.Fluent/Controls/ContentControl.xaml#L1-L16

**Reply by maxkatz6** [reply]
> Or is it not a recommended way to make custom controls?

It's perfectly fine. Just inheriting ContentControl makes it easier. If you don't want to reuse it, you can make completely custom control with ContentPresenter inside of the template - and set content there.
